polkadot/primitives/src/vstaging/mod.rs

```diff
@@ -98,3 +116,873 @@ impl<BlockNumber: Default + From<u32>> Default for SchedulerParams<BlockNumber>
 		}
 	}
 }
+
+/// A type representing the version of the candidate descriptor and internal version number.
+#[derive(PartialEq, Eq, Encode, Decode, Clone, TypeInfo, RuntimeDebug)]
+#[cfg_attr(feature = "std", derive(Hash))]
+pub struct InternalVersion(pub u8);
+
+/// A type representing the version of the candidate descriptor and internal version number.
+/// For `V1`` the internal version number stores the first byte of the `CollatorId`.
+#[derive(PartialEq, Eq, Clone, TypeInfo, RuntimeDebug)]
+#[cfg_attr(feature = "std", derive(Hash))]
+pub enum CandidateDescriptorVersion {
+	/// The old candidate descriptor version.
+	V1,
+	/// Introduced with `CandidateDescriptorV2`
+	V2,
+}
+
+/// A unique descriptor of the candidate receipt.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, TypeInfo, RuntimeDebug)]
+#[cfg_attr(feature = "std", derive(Hash))]
+pub struct CandidateDescriptorV2<H = Hash> {
+	/// The ID of the para this is a candidate for.
+	para_id: ParaId,
+	/// The hash of the relay-chain block this is executed in the context of.
+	relay_parent: H,
+	/// Version field
+	version: InternalVersion,
```

Comment:
I'd add a comment explaining that this is different from the CandidateDescriptorVersion